Add external switch type setting to various Candeo devices

candeodevelopment opened the pull request:
## Proposed change
<!--
  Explain your proposed change below.
-->

Adds specific Quirks for:

Candeo C205 Switch Module
Candeo C-ZB-SM205-2G Switch Module
Candeo C203 Dimmer Module
Candeo C204 Dimmer Module
Candeo C-ZB-DM204 Dimmer Module

## Additional information
<!--
  Please include any additional information that is important to this PR.
  For example, if this PR is a potentially breaking change, mention that here.
  If this PR requires other PRs to be merged in HA Core or other projects, mention that.
  Lastly, if this PR fixes a specific issue, please include "Fixes #xxxx".
-->

We've been providing Quirks to our customers for some time via our own Github repository, we're hoping now to have them officially directly integrated in order to alleviate customers needing to manually add them to their ZHA configuration.

## Checklist
<!--
  Put an 'x' in all boxes that apply.
  Note: You do not need to tick all boxes before creating a PR.
-->

- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [ ] Tests have been added to verify that the new code works

This is our first submission to the project, so hopefully we're doing things correctly.  

Please let us know if not!

candeodevelopment committed:
Initial commit to add support for Candeo devices

Adds specific quirks for:

Candeo C205 Switch Module
Candeo C-ZB-SM205-2G Switch Module
Candeo C203 Dimmer Module
Candeo C204 Dimmer Module
Candeo C-ZB-DM204 Dimmer Module

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

candeodevelopment committed:
Fix pre-commit formatting issues

candeodevelopment committed:
Merge branch 'dev' of https://github.com/candeodevelopment/zha-device-handlers into dev

puddly commented on zhaquirks/candeo/candeo_c205_switch_module_V2.py:
```diff
@@ -0,0 +1,51 @@
+
+(
+    QuirkBuilder(CANDEO, "C205")
+    .replaces(CandeoBasicCluster)
```
I'm noticing a lot of repetition between these quirks. Do the devices differ in functionality or features? If not, you could use `.applies_to()` to share a definition.

candeodevelopment replied:
They are different devices with different features so we'd prefer to have a separate quirk for each of them if possible. We'll likely add additional functionality to each quirk at a later date.

This particular group of quirks just adds the ability to configure the type of switch connected to the input.

puddly commented on zhaquirks/candeo/candeo_c205_switch_module_V2.py:
```diff
@@ -0,0 +1,51 @@
+
+    attr_config = {AttributeDefs.external_switch_type.id: CandeoSwitchType.Toggle}
+
+    async def apply_custom_configuration(self, *args, **kwargs):
```
Is the custom configuration required for the device to function? Or is it resetting it to a default state?

The reason I ask is that we have a separate system for defining binding and attibute reporting. If the attribute write isn't entirely required, if would be best to drop it to remove the need for unit testing this code.

candeodevelopment replied:
This attribute write is to initialise the state of the input to momentary for our dimmer modules and toggle for our switch modules.

puddly replied:
What state is the device in if this attribute isn't written?

candeodevelopment replied:
From the factory the default is momentary for dimmer modules and toggle for switch modules.

The setting doesn't get defaulted after a factory reset of the module though, so we choose to explicitly set it to the default during pairing to ensure the module is in factory default state after pairing.

dhc25 commented:
@TheJulianJES @puddly 
Can we please get an update on this?
Is there any specific reason that is stopping this PR getting merged?
Thank you

TheJulianJES commented:
There's really a lot of duplicated code. It should be combined for the most part.
If different functionality for the quirks is to be added later, it should be split up then.

"External Switch Type" should also be "External switch type" due to HA naming conventions.

A small test should also be added for the custom configuration.

Although, `apply_custom_configuration` will also cause the attribute to be written every time the device is reconfigured, not only when paired. I don't think this is a good idea, as that generally only sets up binding and attribute reporting again and shouldn't change settings on the device. We poll some devices on startup, I don't think we reconfigure them, but I'd still avoid doing this.
The user already has the attribute exposed as a config entity in the UI, so why exactly do we need to reset this every time when using ZHA?

candeodevelopment commented:
> `apply_custom_configuration` will also cause the attribute to be written every time the device is reconfigured

I took inspiration from https://github.com/zigpy/zha-device-handlers/blob/e4844d48549d5a16ed288bb59283a30937ea5591/zhaquirks/philips/__init__.py#L77 

Is there a preferred way to set an initial attribute value only during pairing instead then?

> "External Switch Type" should also be "External switch type" due to HA naming conventions.

OK, we'll get that changed.

> A small test should also be added for the custom configuration.

We're not sure what you mean by this?

> so why exactly do we need to reset this every time when using ZHA

This attribute write is to initialise the state of the input to momentary for our dimmer modules and toggle for our switch modules.

From the factory the default is momentary for dimmer modules and toggle for switch modules.

The setting doesn't get defaulted after a factory reset of the module though, so we choose to explicitly set it to the default during pairing to ensure the module is in factory default state after pairing.

dhc25 commented:
@TheJulianJES Thanks for your feedback.
Can you please let us know how to proceed based on @candeodevelopment feedback above?

puddly commented:
Hi @candeodevelopment, I'm going to reply to you here for visibility.

> The setting doesn't get defaulted after a factory reset of the module though, so we choose to explicitly set it to the default during pairing to ensure the module is in factory default state after pairing.

A factory reset *must* reset the device back to the factory state. If it does not, I would argue it is a firmware bug. We do not support running code on first join in ZHA right now but I'll update you when I add support for this feature soon.

In the mean time, please implement the suggestions @TheJulianJES made earlier.

candeodevelopment committed:
Updated PR inline with discussed requirements

1) Changed string per HA naming conventions
2) Moved common code to module to reduce duplication
3) Removed apply_custom_configuration
4) Replaced individual quirk files with common file for all modules

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

candeodevelopment commented:
Thanks @puddly, we made some changes to the PR, can you let us know if they are OK now please?

1) Changed string per HA naming conventions
2) Moved common code to module to reduce duplication
3) Removed apply_custom_configuration
4) Replaced individual quirk files with common file for all modules

TheJulianJES commented:
I think the PR looks good now.

By the way, for sharing common entities/custom replaced clusters, you can add a base quirk (without matching any models nor adding it to the registry) and then use that base quirk in specific quirks.
I don't think you need it here, but something you might wanna use in the future if you add more device specific options, but have the "external switch type" shared by all quirks, for example.

```python
# base quirk that will be used as a basis for other device specific quirks
base_quirk = (
    QuirkBuilder()
    .sensor(...)  # this will be cloned to both quirks below
    .switch(...)  # and this too
)


(
    base_quirk.clone()  # inherit everything from base quirk (replaced custom clusters, custom entities, ...)
    .applies_to("manufacturer", "model1")
    .switch(...)  # specific to model1
    .add_to_registry()
)


(
    base_quirk.clone()  # inherit everything from base quirk (replaced custom clusters, custom entities, ...)
    .applies_to("manufacturer", "model2")
    .sensor(...)  # specific to model2
    .add_to_registry()
)
```

We'll get docs ready at some point...

candeodevelopment commented:
That's great news @TheJulianJES !

Thanks for the pointers on the base quirk concept, that looks useful.

Since the two quirks here are essentially the same but the second needs a specific endpoint_id, is there was a way to do something like this:

```

base_quirk = (
    QuirkBuilder()
    .enum(with_common_stuff)  
)

(
    base_quirk.clone()  
    .applies_to()
    .enum_from_the_base_quirk(add_some_additional_stuff)
    .add_to_registry()
)
```

Or is that wishful thinking?

TheJulianJES commented on zhaquirks/candeo/candeo_modules_V2.py:
```diff
@@ -0,0 +1,38 @@
+        attribute_name=CandeoBasicCluster.AttributeDefs.external_switch_type.name,
+        enum_class=CandeoSwitchType,
+        cluster_id=CandeoBasicCluster.cluster_id,
+        initially_disabled=False,
```
`initially_disabled` is already `False` by default, so I'll just remove this.
```suggestion
```

TheJulianJES commented on zhaquirks/candeo/candeo_modules_V2.py:
```diff
@@ -0,0 +1,38 @@
+        enum_class=CandeoSwitchType,
+        cluster_id=CandeoBasicCluster.cluster_id,
+        endpoint_id=11,
+        initially_disabled=False,
```
`initially_disabled` is already `False` by default, so I'll just remove this.
```suggestion
```

TheJulianJES committed:
Remove default `initially_disabled` parameter

TheJulianJES reviewed (approved):
Thanks, lgtm!

TheJulianJES commented:
> Since the two quirks here are essentially the same but the second needs a specific endpoint_id, is there was a way to do something like this:

Not at the moment.
